File: legacy/legacy_core_split/packages/logic/src/tick/autoResearch.ts
// Per-tick auto-research dispatch. Lifted from packages/web_ui/src/Synergism.ts
// (`tack` body, the two if-blocks at ~lines 4068-4110 — manual mode + Roomba).
//
// Each branch becomes a single CoreEvent the UI dispatcher translates back
// into the existing DOM-bound side effects:
//   manual:  buyResearch(autoResearch, true, false) + updateResearchAuto
//   cheapest: bounded while-loop reading isResearchUnlocked, buyResearch,
//             researchData[i].maxLevel, updateResearchRoomba — extracted into
//             runRoombaResearchSweep(maxCount) in web_ui.
//
// The two modes are mutually exclusive — only one branch can fire per tick.
// The maxCount calc for Roomba lives here because it's pure math
// (CalcECC is already in logic).

import type { CoreEvent } from '../events/types'
import { CalcECC } from '../mechanics/challenges'

export type AutoResearchMode = 'manual' | 'cheapest'

export interface ProcessAutoResearchTickInput {
  /** player.autoResearchToggle — master gate. When false, no branch fires. */
  autoResearchToggle: boolean
  /** player.autoResearch — current research index; gate (`> 0`) for both
   * modes. Zero means "no research selected" — short-circuit. */
  autoResearch: number
  /** player.autoResearchMode — selects manual vs Roomba branch. The
   * legacy code stores this as a string literal in the savefile. */
  autoResearchMode: AutoResearchMode
  /** Pre-evaluated `roombaResearchEnabled()` from web_ui — `cubeUpgrades[9]
   * === 1 || highestSingularityCount > 10`. Pre-eval'd by the caller so
   * the logic function stays decoupled from cube/singularity state. */
  roombaUnlocked: boolean
  /** player.challengecompletions[14] — feeds `CalcECC('ascension', x)`
   * for the Roomba `maxCount` (`1 + Math.floor(CalcECC(...))`). */
  challengecompletions14: number
}

export interface ProcessAutoResearchTickResult {
  /** Either:
   *   - `[]` — no branch fired (toggle off, no research selected, or
   *           Roomba mode without the unlock)
   *   - `[{ kind: 'auto-research-manual-requested' }]` — manual mode
   *   - `[{ kind: 'auto-research-roomba-requested', maxCount }]` — Roomba
   *
   * Never returns more than one event — the manual + Roomba branches are
   * mutually exclusive on `autoResearchMode`. */
  events: CoreEvent[]
}

/**
 * Per-tick auto-research dispatcher. Mirrors the two if-blocks at the
 * end of the !timeWarp section in `tack` (Synergism.ts).
 *
 * Manual mode (`autoResearchMode === 'manual'`):
 *   if (autoResearchToggle && autoResearch > 0)
 *     → emit `auto-research-manual-requested`
 *
 * Roomba mode (`autoResearchMode === 'cheapest'`):
 *   if (autoResearchToggle && autoResearch > 0 && roombaUnlocked)
 *     → emit `auto-research-roomba-requested` with maxCount =
 *       `1 + floor(CalcECC('ascension', challengecompletions14))`
 *
 * The actual `buyResearch` / `updateResearchAuto` / Roomba while-loop
 * remain in web_ui (DOM-bound) and are invoked by tickEventHandlers on
 * the respective event.
 */
export function processAutoResearchTick (
  input: ProcessAutoResearchTickInput
): ProcessAutoResearchTickResult {
  if (!input.autoResearchToggle || input.autoResearch <= 0) {
    return { events: [] }
  }
  if (input.autoResearchMode === 'manual') {
    return { events: [{ kind: 'auto-research-manual-requested' }] }
  }
  if (input.autoResearchMode === 'cheapest' && input.roombaUnlocked) {
    const maxCount = 1 + Math.floor(CalcECC('ascension', input.challengecompletions14))
    return { events: [{ kind: 'auto-research-roomba-requested', maxCount }] }
  }
  return { events: [] }
}

File: crates/synergismforkd_logic/src/tick/auto_research.rs
//! Per-tick auto-research dispatch. Direct port of
//! `legacy/core_split/packages/logic/src/tick/autoResearch.ts`.
//!
//! Emits one intent event (manual single-slot vs cheapest "Roomba"); the
//! UI tier translates it into the `buyResearch` / Roomba while-loop side
//! effects. The two modes are mutually exclusive.

use smallvec::SmallVec;

use crate::events::CoreEvent;
use crate::mechanics::challenges::{calc_ecc, ChallengeType};
use crate::state::AutoResearchMode;

/// Inputs to [`process_auto_research_tick`].
pub(crate) struct ProcessAutoResearchTickInput {
    /// `player.autoResearchToggle` — master gate.
    pub auto_research_toggle: bool,
    /// `player.autoResearch` — selected research slot; `0` short-circuits.
    pub auto_research_selected: u32,
    /// `player.autoResearchMode` — manual vs cheapest.
    pub auto_research_mode: AutoResearchMode,
    /// Pre-evaluated `roombaResearchEnabled()` — Roomba-mode unlock.
    pub roomba_unlocked: bool,
    /// `player.challengecompletions[14]` — feeds `calc_ecc` for the
    /// Roomba `max_count`.
    pub challengecompletions_14: f64,
}

/// Per-tick auto-research dispatcher. Returns `AutoResearchManualRequested`
/// (manual mode) or `AutoResearchRoombaRequested { max_count }` (cheapest
/// mode, when the Roomba unlock passes); empty when the toggle is off, no
/// research is selected, or Roomba mode lacks its unlock.
pub(crate) fn process_auto_research_tick(
    input: &ProcessAutoResearchTickInput,
) -> SmallVec<[CoreEvent; 1]> {
    let mut events = SmallVec::new();
    if !input.auto_research_toggle || input.auto_research_selected == 0 {
        return events;
    }
    match input.auto_research_mode {
        AutoResearchMode::Manual => events.push(CoreEvent::AutoResearchManualRequested),
        AutoResearchMode::Cheapest => {
            if input.roomba_unlocked {
                let max_count =
                    1 + calc_ecc(ChallengeType::Ascension, input.challengecompletions_14).floor()
                        as u32;
                events.push(CoreEvent::AutoResearchRoombaRequested { max_count });
            }
        }
    }
    events
}

#[cfg(test)]
mod tests {
    use super::*;

    fn input() -> ProcessAutoResearchTickInput {
        ProcessAutoResearchTickInput {
            auto_research_toggle: true,
            auto_research_selected: 5,
            auto_research_mode: AutoResearchMode::Manual,
            roomba_unlocked: false,
            challengecompletions_14: 0.0,
        }
    }

    #[test]
    fn no_event_when_toggle_off() {
        let e = process_auto_research_tick(&ProcessAutoResearchTickInput {
            auto_research_toggle: false,
            ..input()
        });
        assert!(e.is_empty());
    }

    #[test]
    fn no_event_when_no_research_selected() {
        let e = process_auto_research_tick(&ProcessAutoResearchTickInput {
            auto_research_selected: 0,
            ..input()
        });
        assert!(e.is_empty());
    }

    #[test]
    fn manual_mode_emits_manual_request() {
        let e = process_auto_research_tick(&input());
        assert_eq!(e.len(), 1);
        assert!(matches!(e[0], CoreEvent::AutoResearchManualRequested));
    }

    #[test]
    fn cheapest_mode_emits_roomba_when_unlocked() {
        let e = process_auto_research_tick(&ProcessAutoResearchTickInput {
            auto_research_mode: AutoResearchMode::Cheapest,
            roomba_unlocked: true,
            challengecompletions_14: 0.0,
            ..input()
        });
        assert_eq!(e.len(), 1);
        // calc_ecc(Ascension, 0) == 0 → max_count = 1.
        assert!(matches!(
            e[0],
            CoreEvent::AutoResearchRoombaRequested { max_count: 1 }
        ));
    }

    #[test]
    fn cheapest_mode_silent_without_roomba_unlock() {
        let e = process_auto_research_tick(&ProcessAutoResearchTickInput {
            auto_research_mode: AutoResearchMode::Cheapest,
            roomba_unlocked: false,
            ..input()
        });
        assert!(e.is_empty());
    }
}
